src/cpp/src/scheduler.hpp

```diff
@@ -405,6 +435,86 @@ class Scheduler {
             sequence_groups[sequence_group_id]->clear_waiting_sequences();
         }
     }
+
+    size_t _get_available_gpu_memory() {
+        auto device_config = m_cache_manager->get_device_config();
+        auto core = m_cache_manager->get_core();
+        auto device = device_config->get_device();
+        OPENVINO_ASSERT(device.find("GPU") != std::string::npos, "_get_available_gpu_memory() is applicable for GPU only.");
+        auto memory_statistics = core->get_property(device, ov::intel_gpu::memory_statistics);
+        auto device_type = core->get_property(device, ov::device::type);
+
+        // sum up all used device memory
+        std::vector<std::string> device_memory_types = {"cl_mem", "usm_device"};
+        size_t used_device_mem = 0;
+        for (auto mem_type: device_memory_types) {
+            used_device_mem += memory_statistics[mem_type];
+        }
+
+        if (device_type == ov::device::Type::INTEGRATED) {
+            used_device_mem += memory_statistics["usm_host"];
+        }
+
+        // there could be unaccounted extra memory reserved by kernels, kept
+        // in memory pools, etc
+        // therefore, add a threshold to account for this
+        float used_memory_threshold = 1.1;
+        used_device_mem *= used_memory_threshold;
+
+        // total device memory in bytes
+        auto total_device_memory = core->get_property(device, ov::intel_gpu::device_total_mem_size);
+
+        return total_device_memory - used_device_mem;
+    }
+
+    void _initialize_cache(const std::vector<SequenceGroup::Ptr>& sequence_groups) {
+        size_t blocks_sum = 0;
+        for (auto idx = 0; idx < sequence_groups.size(); idx++) {
+            auto seq_length = sequence_groups[idx]->get_prompt_len() * m_kv_blocks_initial_multiplier;
+            auto gen_config = sequence_groups[idx]->get_sampling_parameters();
+            seq_length = std::min(seq_length, sequence_groups[idx]->get_prompt_len() + gen_config.get_max_new_tokens(sequence_groups[idx]->get_prompt_len()));
+            size_t blocks_num = std::ceil((float)seq_length / m_block_manager.get_block_size());
+            if (gen_config.do_sample && gen_config.is_beam_search()) {
```

Comment:
```suggestion
            if (gen_config.is_beam_search()) {
```
currently, we don't support do_sample together with beam search. Those are 2 separate sampling strategies